### DISCUSSION TITLE
Websocket based live-preview

### DISCUSSION THREAD

**Discussion by LasseRosenow** [directus/directus#18792] [Feature Requests]
### Summary

Directus just got a "live-preview" and I am really happy about it, but it is not as live as it could be.
Directus live-preview is only live in terms of when the `save`  button is clicked.

Other Headless-CMSs use Websockets instead. e.g. `Storyblok`.
This way the live-preview could truly be live and every written letter would immediately show.
So we can see how our updated page looks like before saving it.

### Motivation

- Seeing the result of our edits without having to continuously press `save` buttons and waiting for a page refresh saves a lot of time.
- Also things like scroll-position etc. would always stay correct.

### Detailed Design

1. The client website could detect that is in preview via some flag in the `url` 
2. It could for example then do a graphql `subscription` instead of a `graphql-request`.
3. The graphql-subscription maybe has a `preview` flag somewhere
4. When content gets changed without pressing the `save` button, `graphql-subscriptions` that use the `preview` flag get the new input data
5. The client rerenders the page

### Must Have:
- WebSockets for Graphql, REST,  etc.
- Ability to disable the reload of the live-preview iframe (preferably via ENV-variable)

### Should Have:
- Some way to tell the Directus CMS that a `preview` content is requested

### Drawbacks

- Implementation effort, but I think it is really worth it.
- Developer implementation effort
  - But it is optional
  - It can be massively improved with a good Directus SDK that takes care of most of the boilerplate (See Storyblok)

### Alternatives

Just reload the iframe like how it is implemented right now.

### Adoption Strategy

This is non-breaking.
Developers can opt in to it because clients need to implement it

**Comment by rijkvanzanten** [comment]
Yes yes! My goal is to have https://github.com/directus/directus/discussions/2975 first, then allow you to subscribe to this "/preview" endpoint that includes the unsaved changes, and then have a real-time live preview 😄

**Reply by LasseRosenow** [reply]
Amazing amazing amazing :) ❤️❤️ 
Thank you for your hard work, I really love directus more and more every day I use it :)

**Comment by Natetronn** [comment]
I would love to see this FR implemented for my client's/user's sake! It's such a cool feature.

**Comment by yfi** [comment]
FWIW, the live preview implementation in Payload 2 is quite clever and simple.

The admin panel sends data to the preview endpoint iframe via window.postMessage triggered by field input events. The nicest part about this is that it allows you to preview the changes in realtime prior to saving the document. Also, the preview endpoint can load with initial date pulled from the database prior to establishing the inter window connection so you can share the preview url with saved content.

I realize this doesn't work for all types of frontends but it was an absolute breeze to implement in my sveltekit app. I just needed to change the data source in my load function - the rendering components did not require any changes ( whereas in storyblok you need to inject their sdk into each component, at least last time i used it. )

Perhaps this idea can be implemented as an extension. I could take a crack at it.
